tools/im2rec.py

```diff
@@ -1,238 +1,265 @@
+        sep_test = int(chunk_size * args.test_ratio)
+        write_list(args.prefix + str_chunk + '_test.lst', chunk[:sep_test])
+        write_list(args.prefix + str_chunk + '_train.lst', chunk[sep_test:sep_test + sep])
+        write_list(args.prefix + str_chunk + '_val.lst', chunk[sep_test + sep:])
+
+
+def read_list(path_in):
+    image_list = []
+    with open(path_in) as fin:
+        for line in fin.readlines():
+            line = [i.strip() for i in line.strip().split('\t')]
+            item = [int(line[0])] + [line[-1]] + [int(i) for i in line[1:-1]]
+            image_list.append(item)
+    return image_list
+
+
+# Changed the original function write_record cause the multiprocessing must be in the __main__ process in Windows, otherwise the Pickle
+# Error would happen
+
+def image_encode(args, item, q_out):
+    # move the coler modes here
+    color_modes = {-1: cv2.IMREAD_UNCHANGED,
+                   0: cv2.IMREAD_GRAYSCALE,
+                   1: cv2.IMREAD_COLOR}
+
+    # cause the content of make_list, here get the parient directory of data root
+    pari_path = os.path.relpath(os.path.join(args.root, '..'), '.')
+    try:
+        # change the file path by join pari_path and item[1]
+        img = cv2.imread(os.path.join(pari_path, item[1]), color_modes[args.color])
+    except:
+        print 'imread error:', item[1]
+        return
+    if img is None:
+        print 'read none error:', item[1]
+        return
+    if args.center_crop:
+        if img.shape[0] > img.shape[1]:
+            margin = (img.shape[0] - img.shape[1]) / 2;
+            img = img[margin:margin + img.shape[1], :]
+        else:
+            margin = (img.shape[1] - img.shape[0]) / 2;
+            img = img[:, margin:margin + img.shape[0]]
+    if args.resize:
+        if img.shape[0] > img.shape[1]:
+            newsize = (args.resize, img.shape[0] * args.resize / img.shape[1])
+        else:
+            newsize = (img.shape[1] * args.resize / img.shape[0], args.resize)
+        img = cv2.resize(img, newsize)
+    header = mx.recordio.IRHeader(0, item[2], item[0], 0)
+
+    try:
+        s = mx.recordio.pack_img(header, img, quality=args.quality, img_fmt=args.encoding)
+        q_out.put(('data', s, item))
+    except:
+        print 'pack_img error:', item[1]
+        return
+
+# the original read_worker in write_record, add argument args
+def read_worker(args, q_in, q_out):
+    while not q_in.empty():
+        item = q_in.get()
+        image_encode(args, item, q_out)
+
+# the write_worker 
+def write_worker(q_out, fname, saving_folder):
+    pre_time = time.time()
+    sink = []
+    fname_rec = fname[:fname.rfind('.')]
+    
+    # remove the change directory operation, change the write record operation controled by write function it self 
+    rec_file = os.path.join(saving_folder, fname_rec + '.rec')
+    record = mx.recordio.MXRecordIO(rec_file, 'w')
+    
+    while True:
+        stat, s, item = q_out.get()
+        if stat == 'finish':

```

Comment:
# .lst file is already exists so this action is not needed
 +
 +            # lst_file = os.path.join(saving_folder, fname_rec + '.lst')
 +            # write_list(lst_file, sink)